File: server/src/main/java/com/uiptv/service/BookmarkResolver.java
package com.uiptv.service;

import com.uiptv.model.Account;
import com.uiptv.model.Bookmark;
import com.uiptv.model.Channel;
import com.uiptv.shared.Episode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.uiptv.util.StringUtils.isBlank;
import static com.uiptv.util.StringUtils.isNotBlank;

public class BookmarkResolver {

    public ResolutionContext prepare(List<Bookmark> bookmarks) {
        Map<String, Account> accountByName = AccountService.getInstance().getAll();
        Map<String, BookmarkRenderData> renderDataByBookmarkId = preloadRenderData(bookmarks);
        Map<String, Channel> channelByAccountAndChannel = preloadFallbackChannels(bookmarks, accountByName, renderDataByBookmarkId);
        return new ResolutionContext(accountByName, channelByAccountAndChannel, renderDataByBookmarkId);
    }

    public List<ResolvedBookmark> resolveBookmarks(List<Bookmark> bookmarks) {
        ResolutionContext context = prepare(bookmarks);
        List<ResolvedBookmark> resolved = new ArrayList<>();
        if (bookmarks == null) {
            return resolved;
        }
        for (Bookmark bookmark : bookmarks) {
            resolved.add(resolveBookmark(bookmark, context));
        }
        return resolved;
    }

    public ResolvedBookmark resolveBookmark(Bookmark bookmark, ResolutionContext context) {
        Account account = context.accountByName.get(bookmark.getAccountName());
        BookmarkRenderData renderData = context.renderDataByBookmarkId.getOrDefault(bookmarkKey(bookmark), BookmarkRenderData.fromBookmark(bookmark));
        if (needsChannelFallback(renderData)) {
            mergeRenderData(renderData, lookupFallbackChannel(account, bookmark.getChannelId(), context.channelByAccountAndChannel));
        }

        Account.AccountAction accountAction = bookmark.getAccountAction();
        if (accountAction == null) {
            accountAction = account != null ? account.getAction() : Account.AccountAction.itv;
        }
        ResolvedBookmark resolved = new ResolvedBookmark(bookmark, account, accountAction, renderData);
        resolved.applyToBookmark();
        return resolved;
    }

    private BookmarkRenderData resolveBookmarkRenderData(Bookmark bookmark) {
        BookmarkRenderData renderData = BookmarkRenderData.fromBookmark(bookmark);
        mergeRenderData(renderData, resolveBookmarkChannelSnapshot(bookmark));
        return renderData;
    }

    private Map<String, BookmarkRenderData> preloadRenderData(List<Bookmark> bookmarks) {
        Map<String, BookmarkRenderData> renderDataByBookmarkId = new HashMap<>();
        if (bookmarks == null || bookmarks.isEmpty()) {
            return renderDataByBookmarkId;
        }
        for (Bookmark bookmark : bookmarks) {
            if (bookmark == null) {
                continue;
            }
            renderDataByBookmarkId.put(bookmarkKey(bookmark), resolveBookmarkRenderData(bookmark));
        }
        return renderDataByBookmarkId;
    }

    private Channel resolveBookmarkChannelSnapshot(Bookmark bookmark) {
        if (bookmark == null) {
            return null;
        }
        if (isNotBlank(bookmark.getChannelJson())) {
            Channel channel = Channel.fromJson(bookmark.getChannelJson());
            if (channel != null) {
                return channel;
            }
        }
        if (isNotBlank(bookmark.getVodJson())) {
            Channel channel = Channel.fromJson(bookmark.getVodJson());
            if (channel != null) {
                return channel;
            }
        }
        if (isNotBlank(bookmark.getSeriesJson())) {
            Episode episode = Episode.fromJson(bookmark.getSeriesJson());
            if (episode != null) {
                Channel channel = new Channel();
                channel.setLogo(episode.getInfo() != null ? episode.getInfo().getMovieImage() : "");
                return channel;
            }
        }
        return null;
    }

    private boolean needsChannelFallback(BookmarkRenderData renderData) {
        return renderData == null
                || isBlank(renderData.logo)
                || isBlank(renderData.drmType)
                || isBlank(renderData.drmLicenseUrl)
                || isBlank(renderData.clearKeysJson)
                || isBlank(renderData.inputstreamaddon)
                || isBlank(renderData.manifestType);
    }

    private void mergeRenderData(BookmarkRenderData target, Channel channel) {
        if (target == null || channel == null) {
            return;
        }
        if (isBlank(target.logo)) target.logo = channel.getLogo();
        if (isBlank(target.drmType)) target.drmType = channel.getDrmType();
        if (isBlank(target.drmLicenseUrl)) target.drmLicenseUrl = channel.getDrmLicenseUrl();
        if (isBlank(target.clearKeysJson)) target.clearKeysJson = channel.getClearKeysJson();
        if (isBlank(target.inputstreamaddon)) target.inputstreamaddon = channel.getInputstreamaddon();
        if (isBlank(target.manifestType)) target.manifestType = channel.getManifestType();
    }

    private Map<String, Channel> preloadFallbackChannels(List<Bookmark> bookmarks,
                                                         Map<String, Account> accountByName,
                                                         Map<String, BookmarkRenderData> renderDataByBookmarkId) {
        Map<String, Channel> channelByAccountAndChannel = new HashMap<>();
        if (bookmarks == null || bookmarks.isEmpty() || accountByName == null || accountByName.isEmpty()) {
            return channelByAccountAndChannel;
        }

        Map<String, List<String>> requestedChannelIdsByAccountId = collectFallbackChannelIds(bookmarks, accountByName, renderDataByBookmarkId);
        loadFallbackChannelsIntoCache(requestedChannelIdsByAccountId, channelByAccountAndChannel);

        return channelByAccountAndChannel;
    }

    private Map<String, List<String>> collectFallbackChannelIds(List<Bookmark> bookmarks,
                                                                Map<String, Account> accountByName,
                                                                Map<String, BookmarkRenderData> renderDataByBookmarkId) {
        Map<String, List<String>> requestedChannelIdsByAccountId = new HashMap<>();
        for (Bookmark bookmark : bookmarks) {
            if (!requiresFallbackLookup(bookmark, accountByName, renderDataByBookmarkId)) {
                continue;
            }
            Account account = accountByName.get(bookmark.getAccountName());
            requestedChannelIdsByAccountId
                    .computeIfAbsent(account.getDbId(), ignored -> new ArrayList<>())
                    .add(bookmark.getChannelId());
        }
        return requestedChannelIdsByAccountId;
    }

    private boolean requiresFallbackLookup(Bookmark bookmark,
                                           Map<String, Account> accountByName,
                                           Map<String, BookmarkRenderData> renderDataByBookmarkId) {
        if (bookmark == null) {
            return false;
        }
        BookmarkRenderData renderData = renderDataByBookmarkId.get(bookmarkKey(bookmark));
        if (!needsChannelFallback(renderData)) {
            return false;
        }
        Account account = accountByName.get(bookmark.getAccountName());
        return account != null && isNotBlank(account.getDbId()) && isNotBlank(bookmark.getChannelId());
    }

    private void loadFallbackChannelsIntoCache(Map<String, List<String>> requestedChannelIdsByAccountId,
                                               Map<String, Channel> channelByAccountAndChannel) {
        for (Map.Entry<String, List<String>> entry : requestedChannelIdsByAccountId.entrySet()) {
            List<Channel> channels = ChannelService.getInstance().getChannelsByChannelIdsAndAccount(entry.getValue(), entry.getKey(), false);
            cacheChannelsByAccountAndId(entry.getKey(), channels, channelByAccountAndChannel);
        }
    }

    private void cacheChannelsByAccountAndId(String accountId, List<Channel> channels, Map<String, Channel> channelByAccountAndChannel) {
        for (Channel channel : channels) {
            if (channel == null || isBlank(channel.getChannelId())) {
                continue;
            }
            channelByAccountAndChannel.put(accountId + "|" + channel.getChannelId(), channel);
        }
    }

    private Channel lookupFallbackChannel(Account account, String channelId, Map<String, Channel> channelByAccountAndChannel) {
        if (account == null || isBlank(account.getDbId()) || isBlank(channelId)) {
            return null;
        }
        String key = account.getDbId() + "|" + channelId;
        if (channelByAccountAndChannel.containsKey(key)) {
            return channelByAccountAndChannel.get(key);
        }
        Channel channel = null;
        try {
            channel = ChannelService.getInstance().getChannelByChannelIdAndAccount(channelId, account.getDbId(), false);
        } catch (Exception _) {
            // Best-effort fallback only. Resolver should not fail on auxiliary channel lookup.
        }
        channelByAccountAndChannel.put(key, channel);
        return channel;
    }

    private String bookmarkKey(Bookmark bookmark) {
        if (bookmark == null) {
            return "";
        }
        if (isNotBlank(bookmark.getDbId())) {
            return bookmark.getDbId();
        }
        return (bookmark.getAccountName() == null ? "" : bookmark.getAccountName()) + "|"
                + (bookmark.getChannelId() == null ? "" : bookmark.getChannelId()) + "|"
                + (bookmark.getChannelName() == null ? "" : bookmark.getChannelName());
    }

    public static final class ResolutionContext {
        private final Map<String, Account> accountByName;
        private final Map<String, Channel> channelByAccountAndChannel;
        private final Map<String, BookmarkRenderData> renderDataByBookmarkId;

        private ResolutionContext(Map<String, Account> accountByName,
                                  Map<String, Channel> channelByAccountAndChannel,
                                  Map<String, BookmarkRenderData> renderDataByBookmarkId) {
            this.accountByName = accountByName == null ? Map.of() : accountByName;
            this.channelByAccountAndChannel = channelByAccountAndChannel == null ? Map.of() : channelByAccountAndChannel;
            this.renderDataByBookmarkId = renderDataByBookmarkId == null ? Map.of() : renderDataByBookmarkId;
        }
    }

    public static final class ResolvedBookmark {
        private final Bookmark bookmark;
        private final Account account;
        private final Account.AccountAction accountAction;
        private final BookmarkRenderData renderData;

        private ResolvedBookmark(Bookmark bookmark, Account account, Account.AccountAction accountAction, BookmarkRenderData renderData) {
            this.bookmark = bookmark;
            this.account = account;
            this.accountAction = accountAction;
            this.renderData = renderData != null ? renderData : new BookmarkRenderData();
        }

        public Bookmark getBookmark() {
            return bookmark;
        }

        public Account getAccount() {
            return account;
        }

        public Account.AccountAction getAccountAction() {
            return accountAction;
        }

        public String getLogo() {
            return renderData.logo;
        }

        public String getDrmType() {
            return renderData.drmType;
        }

        public String getDrmLicenseUrl() {
            return renderData.drmLicenseUrl;
        }

        public String getClearKeysJson() {
            return renderData.clearKeysJson;
        }

        public String getInputstreamaddon() {
            return renderData.inputstreamaddon;
        }

        public String getManifestType() {
            return renderData.manifestType;
        }

        public void applyToBookmark() {
            if (bookmark == null) {
                return;
            }
            if (bookmark.getAccountAction() == null && accountAction != null) {
                bookmark.setAccountAction(accountAction);
            }
            bookmark.setLogo(renderData.logo);
            bookmark.setDrmType(renderData.drmType);
            bookmark.setDrmLicenseUrl(renderData.drmLicenseUrl);
            bookmark.setClearKeysJson(renderData.clearKeysJson);
            bookmark.setInputstreamaddon(renderData.inputstreamaddon);
            bookmark.setManifestType(renderData.manifestType);
        }
    }

    private static final class BookmarkRenderData {
        private String logo;
        private String drmType;
        private String drmLicenseUrl;
        private String clearKeysJson;
        private String inputstreamaddon;
        private String manifestType;

        private static BookmarkRenderData fromBookmark(Bookmark bookmark) {
            BookmarkRenderData data = new BookmarkRenderData();
            if (bookmark == null) {
                return data;
            }
            data.logo = bookmark.getLogo();
            data.drmType = bookmark.getDrmType();
            data.drmLicenseUrl = bookmark.getDrmLicenseUrl();
            data.clearKeysJson = bookmark.getClearKeysJson();
            data.inputstreamaddon = bookmark.getInputstreamaddon();
            data.manifestType = bookmark.getManifestType();
            return data;
        }
    }
}

File: server/src/main/kotlin/com/uiptv/service/BookmarkResolver.kt
package com.uiptv.service

import com.uiptv.model.Account
import com.uiptv.model.Bookmark
import com.uiptv.model.Channel
import com.uiptv.shared.Episode
import com.uiptv.util.StringUtils.isBlank
import com.uiptv.util.StringUtils.isNotBlank

class BookmarkResolver(
    private val accountServiceProvider: () -> AccountService = { AccountService },
    private val channelServiceProvider: () -> ChannelService
) {
    fun prepare(bookmarks: List<Bookmark>?): ResolutionContext {
        val accountByName = accountServiceProvider.invoke().getAll()
        val renderDataByBookmarkId = preloadRenderData(bookmarks)
        val channelByAccountAndChannel = preloadFallbackChannels(bookmarks, accountByName, renderDataByBookmarkId)
        return ResolutionContext(accountByName, channelByAccountAndChannel, renderDataByBookmarkId)
    }

    fun resolveBookmarks(bookmarks: List<Bookmark>?): List<ResolvedBookmark> {
        val context = prepare(bookmarks)
        if (bookmarks.isNullOrEmpty()) {
            return emptyList()
        }
        return bookmarks.map { resolveBookmark(it, context) }
    }

    fun resolveBookmark(bookmark: Bookmark, context: ResolutionContext): ResolvedBookmark {
        val account = context.accountByName[bookmark.accountName]
        val renderData = context.renderDataByBookmarkId[bookmarkKey(bookmark)] ?: BookmarkRenderData.fromBookmark(bookmark)
        if (needsChannelFallback(renderData)) {
            mergeRenderData(renderData, lookupFallbackChannel(account, bookmark.channelId, context.channelByAccountAndChannel))
        }

        val accountAction = bookmark.accountAction ?: account?.action ?: Account.AccountAction.itv
        return ResolvedBookmark(bookmark, account, accountAction, renderData).apply {
            applyToBookmark()
        }
    }

    private fun resolveBookmarkRenderData(bookmark: Bookmark?): BookmarkRenderData {
        val renderData = BookmarkRenderData.fromBookmark(bookmark)
        mergeRenderData(renderData, resolveBookmarkChannelSnapshot(bookmark))
        return renderData
    }

    private fun preloadRenderData(bookmarks: List<Bookmark>?): Map<String, BookmarkRenderData> {
        if (bookmarks.isNullOrEmpty()) {
            return emptyMap()
        }
        val renderDataByBookmarkId = LinkedHashMap<String, BookmarkRenderData>()
        bookmarks.forEach { bookmark ->
            renderDataByBookmarkId[bookmarkKey(bookmark)] = resolveBookmarkRenderData(bookmark)
        }
        return renderDataByBookmarkId
    }

    private fun resolveBookmarkChannelSnapshot(bookmark: Bookmark?): Channel? {
        if (bookmark == null) {
            return null
        }
        if (isNotBlank(bookmark.channelJson)) {
            Channel.fromJson(bookmark.channelJson.orEmpty())?.let { return it }
        }
        if (isNotBlank(bookmark.vodJson)) {
            Channel.fromJson(bookmark.vodJson.orEmpty())?.let { return it }
        }
        if (isNotBlank(bookmark.seriesJson)) {
            val episode = Episode.fromJson(bookmark.seriesJson)
            if (episode != null) {
                return Channel().apply {
                    logo = episode.info?.movieImage ?: ""
                }
            }
        }
        return null
    }

    private fun needsChannelFallback(renderData: BookmarkRenderData?): Boolean =
        renderData == null ||
            isBlank(renderData.logo) ||
            isBlank(renderData.drmType) ||
            isBlank(renderData.drmLicenseUrl) ||
            isBlank(renderData.clearKeysJson) ||
            isBlank(renderData.inputstreamaddon) ||
            isBlank(renderData.manifestType)

    private fun mergeRenderData(target: BookmarkRenderData?, channel: Channel?) {
        if (target == null || channel == null) {
            return
        }
        if (isBlank(target.logo)) target.logo = channel.logo
        if (isBlank(target.drmType)) target.drmType = channel.drmType
        if (isBlank(target.drmLicenseUrl)) target.drmLicenseUrl = channel.drmLicenseUrl
        if (isBlank(target.clearKeysJson)) target.clearKeysJson = channel.clearKeysJson
        if (isBlank(target.inputstreamaddon)) target.inputstreamaddon = channel.inputstreamaddon
        if (isBlank(target.manifestType)) target.manifestType = channel.manifestType
    }

    private fun preloadFallbackChannels(
        bookmarks: List<Bookmark>?,
        accountByName: Map<String, Account>,
        renderDataByBookmarkId: Map<String, BookmarkRenderData>
    ): MutableMap<String, Channel> {
        if (bookmarks.isNullOrEmpty() || accountByName.isEmpty()) {
            return LinkedHashMap()
        }
        val channelByAccountAndChannel = LinkedHashMap<String, Channel>()
        val requestedChannelIdsByAccountId = collectFallbackChannelIds(bookmarks, accountByName, renderDataByBookmarkId)
        loadFallbackChannelsIntoCache(requestedChannelIdsByAccountId, channelByAccountAndChannel)
        return channelByAccountAndChannel
    }

    private fun collectFallbackChannelIds(
        bookmarks: List<Bookmark>,
        accountByName: Map<String, Account>,
        renderDataByBookmarkId: Map<String, BookmarkRenderData>
    ): Map<String, List<String>> {
        val requestedChannelIdsByAccountId = LinkedHashMap<String, MutableList<String>>()
        bookmarks.forEach { bookmark ->
            if (!requiresFallbackLookup(bookmark, accountByName, renderDataByBookmarkId)) {
                return@forEach
            }
            val account = accountByName[bookmark.accountName] ?: return@forEach
            requestedChannelIdsByAccountId.computeIfAbsent(account.dbId.orEmpty()) { ArrayList() }
                .add(bookmark.channelId.orEmpty())
        }
        return requestedChannelIdsByAccountId
    }

    private fun requiresFallbackLookup(
        bookmark: Bookmark?,
        accountByName: Map<String, Account>,
        renderDataByBookmarkId: Map<String, BookmarkRenderData>
    ): Boolean {
        if (bookmark == null) {
            return false
        }
        val renderData = renderDataByBookmarkId[bookmarkKey(bookmark)]
        if (!needsChannelFallback(renderData)) {
            return false
        }
        val account = accountByName[bookmark.accountName]
        return account != null && isNotBlank(account.dbId) && isNotBlank(bookmark.channelId)
    }

    private fun loadFallbackChannelsIntoCache(
        requestedChannelIdsByAccountId: Map<String, List<String>>,
        channelByAccountAndChannel: MutableMap<String, Channel>
    ) {
        requestedChannelIdsByAccountId.forEach { (accountId, channelIds) ->
            val channels = channelServiceProvider.invoke().getChannelsByChannelIdsAndAccount(channelIds, accountId, false)
            cacheChannelsByAccountAndId(accountId, channels, channelByAccountAndChannel)
        }
    }

    private fun cacheChannelsByAccountAndId(accountId: String, channels: List<Channel>, channelByAccountAndChannel: MutableMap<String, Channel>) {
        channels.forEach { channel ->
            if (isBlank(channel.channelId)) {
                return@forEach
            }
            channelByAccountAndChannel["$accountId|${channel.channelId}"] = channel
        }
    }

    private fun lookupFallbackChannel(account: Account?, channelId: String?, channelByAccountAndChannel: MutableMap<String, Channel>): Channel? {
        if (account == null || isBlank(account.dbId) || isBlank(channelId)) {
            return null
        }
        val key = account.dbId.orEmpty() + "|" + channelId
        if (channelByAccountAndChannel.containsKey(key)) {
            return channelByAccountAndChannel[key]
        }
        val channel = try {
            channelServiceProvider.invoke().getChannelByChannelIdAndAccount(channelId.orEmpty(), account.dbId.orEmpty(), false)
        } catch (_: Exception) {
            null
        }
        if (channel != null) {
            channelByAccountAndChannel[key] = channel
        }
        return channel
    }

    private fun bookmarkKey(bookmark: Bookmark?): String {
        if (bookmark == null) {
            return ""
        }
        if (isNotBlank(bookmark.dbId)) {
            return bookmark.dbId.orEmpty()
        }
        return bookmark.accountName.orEmpty() + "|" +
            bookmark.channelId.orEmpty() + "|" +
            bookmark.channelName.orEmpty()
    }

    class ResolutionContext(
        val accountByName: Map<String, Account> = emptyMap(),
        val channelByAccountAndChannel: MutableMap<String, Channel> = LinkedHashMap(),
        val renderDataByBookmarkId: Map<String, BookmarkRenderData> = emptyMap()
    )

    class ResolvedBookmark(
        val bookmark: Bookmark?,
        val account: Account?,
        val accountAction: Account.AccountAction?,
        private val renderData: BookmarkRenderData = BookmarkRenderData()
    ) {
        fun getLogo(): String? = renderData.logo
        fun getDrmType(): String? = renderData.drmType
        fun getDrmLicenseUrl(): String? = renderData.drmLicenseUrl
        fun getClearKeysJson(): String? = renderData.clearKeysJson
        fun getInputstreamaddon(): String? = renderData.inputstreamaddon
        fun getManifestType(): String? = renderData.manifestType

        fun applyToBookmark() {
            val currentBookmark = bookmark ?: return
            if (currentBookmark.accountAction == null && accountAction != null) {
                currentBookmark.accountAction = accountAction
            }
            currentBookmark.logo = renderData.logo
            currentBookmark.drmType = renderData.drmType
            currentBookmark.drmLicenseUrl = renderData.drmLicenseUrl
            currentBookmark.clearKeysJson = renderData.clearKeysJson
            currentBookmark.inputstreamaddon = renderData.inputstreamaddon
            currentBookmark.manifestType = renderData.manifestType
        }
    }

    class BookmarkRenderData {
        var logo: String? = null
        var drmType: String? = null
        var drmLicenseUrl: String? = null
        var clearKeysJson: String? = null
        var inputstreamaddon: String? = null
        var manifestType: String? = null

        companion object {
            @JvmStatic
            fun fromBookmark(bookmark: Bookmark?): BookmarkRenderData {
                val data = BookmarkRenderData()
                if (bookmark == null) {
                    return data
                }
                data.logo = bookmark.logo
                data.drmType = bookmark.drmType
                data.drmLicenseUrl = bookmark.drmLicenseUrl
                data.clearKeysJson = bookmark.clearKeysJson
                data.inputstreamaddon = bookmark.inputstreamaddon
                data.manifestType = bookmark.manifestType
                return data
            }
        }
    }
}
